lib/collision_detector_2.cpp

```diff
@@ -0,0 +1,276 @@
+            }
+        }
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::Y, Direction::NEGATIVE);
+            if (collision.space_left >= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+        {
+            auto collision = this->detectContinuousCollision(entity_id, Axis::Y, Direction::POSITIVE);
+            if (collision.space_left <= 0 && collision.opposite_id) {
+                collisions.push_back(Collision(*collision.opposite_id, false));
+            }
+        }
+    }
+    return collisions;
+}
+
+bool CollisionDetector2::findDisplacementException(std::set<std::string> tags_1, std::set<std::string> tags_2) const {
+    for (const std::string& tag_1 : tags_1) {
+        if (is_trigger_tag_exceptions.count(tag_1)) {
+            for (const std::string& tag_2 : tags_2) {
+                if (is_trigger_tag_exceptions.at(tag_1).count(tag_2)) {
+                    return true;
+                }
+            }
+        }
+    }
+    for (const std::string& tag_2 : tags_2) {
+        if (is_trigger_tag_exceptions.count(tag_2)) {
+            for (const std::string& tag_1 : tags_1) {
+                if (is_trigger_tag_exceptions.at(tag_2).count(tag_1)) {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
+
+
+CollisionDetector2CacheInfo CollisionDetector2::getCacheInfo() const {
+    return cache_info;
+}
+
+void CollisionDetector2::invalidateCache() {
+    cache_info = CollisionDetector2CacheInfo();
+    discrete_cache.clear();
+    continuous_cache.clear();
+}
+
+std::vector<DiscreteCollision> CollisionDetector2::detectDiscreteCollision(int entity_id) {
+    if (discrete_cache.count(entity_id)) {
+        ++cache_info.discrete_cache_hits;
+        return discrete_cache.at(entity_id);
+    }
+
+    auto entity_collider = em.getComponent<RectangleColliderComponent>(entity_id);
+    auto [ entity_position, entity_scale ] = em.getAbsoluteTransform(entity_id);
+    auto entity_parent = em.getParent(entity_id);
+    auto entity_children = em.getChildren(entity_id);
+
+    auto opposite_entities_with_collider = em.getEntitiesByComponent<RectangleColliderComponent>();
+
+    std::vector<DiscreteCollision> collisions;
+
+    for (auto& [ opposite_id, opposite_collider] : opposite_entities_with_collider) {
+        // Can't collide with yourself
+        if (opposite_id == entity_id) continue;
+        auto [ opposite_position, opposite_scale ] = em.getAbsoluteTransform(opposite_id);
+        // Can't collide with family
+        if (entity_parent && *entity_parent == opposite_id) continue;
+        if (entity_children.count(opposite_id)) continue;
+
+        bool is_trigger = false;
+        if (entity_collider->is_trigger || opposite_collider->is_trigger) {
+            // If there are no is_trigger_exceptions for these colliding entities,
```

Comment:
Are these like actual exceptions, so an egg cannot collide with chickens for instance? Not a stackoverflowexception?

Reply:
See comment on arena